File: day15/main.go
package main

import (
	"bufio"
	"container/heap"
	"fmt"
	"log"
	"os"
	"sort"
)

const (
	puzzleInput = "input.txt"
)

func copyMap(m map[Pos]*Entity) map[Pos]*Entity {
	m2 := map[Pos]*Entity{}
	for k, v := range m {
		a := *v
		m2[k] = &a
	}
	return m2
}

type (
	Pos struct {
		x, y int
	}

	PosList []Pos
)

func (s PosList) Len() int {
	return len(s)
}
func (s PosList) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}
func (s PosList) Less(i, j int) bool {
	return s[i].Less(s[j])
}

func (p Pos) Less(other Pos) bool {
	if p.y == other.y {
		return p.x < other.x
	}
	return p.y < other.y
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func (p Pos) Manhattan(other Pos) int {
	return abs(other.x-p.x) + abs(other.y-p.y)
}

type (
	Entity struct {
		elf    bool
		pos    Pos
		health int
		attack int
		debug  bool
	}

	EntityByPos []*Entity
)

func NewEntity(pos Pos, elf bool) *Entity {
	return &Entity{
		pos:    pos,
		elf:    elf,
		health: 200,
		attack: 3,
	}
}

func (e *Entity) Dead() bool {
	return e.health < 1
}
func (e *Entity) Hit(damage int) bool {
	e.health -= damage
	return e.Dead()
}
func (e *Entity) Attack(other *Entity) bool {
	return other.Hit(e.attack)
}

func (s EntityByPos) Len() int {
	return len(s)
}
func (s EntityByPos) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}
func (s EntityByPos) Less(i, j int) bool {
	return s[i].pos.Less(s[j].pos)
}

type (
	Board [][]byte

	Game struct {
		elfs    map[Pos]*Entity
		goblins map[Pos]*Entity
		board   Board
		elfDied bool
	}
)

func NewGame(elfs, goblins map[Pos]*Entity, board [][]byte) *Game {
	return &Game{
		elfs:    elfs,
		goblins: goblins,
		board:   Board(board),
		elfDied: false,
	}
}

func (g *Game) IsFree(pos Pos) bool {
	_, hasElf := g.elfs[pos]
	_, hasGoblin := g.goblins[pos]
	return !hasElf && !hasGoblin && g.board[pos.y][pos.x] == byte('.')
}

func (g *Game) AdjacentFree(pos Pos) []Pos {
	k := []Pos{}
	i := Pos{
		x: pos.x,
		y: pos.y - 1,
	}
	if g.IsFree(i) {
		k = append(k, i)
	}
	i = Pos{
		x: pos.x - 1,
		y: pos.y,
	}
	if g.IsFree(i) {
		k = append(k, i)
	}
	i = Pos{
		x: pos.x + 1,
		y: pos.y,
	}
	if g.IsFree(i) {
		k = append(k, i)
	}
	i = Pos{
		x: pos.x,
		y: pos.y + 1,
	}
	if g.IsFree(i) {
		k = append(k, i)
	}
	return k
}

func (g *Game) IsEnemy(pos Pos, elf bool) bool {
	if elf {
		_, ok := g.goblins[pos]
		return ok
	}
	_, ok := g.elfs[pos]
	return ok
}

func (g *Game) AdjacentEnemy(pos Pos, elf bool) []Pos {
	k := []Pos{}
	i := Pos{
		x: pos.x,
		y: pos.y - 1,
	}
	if g.IsEnemy(i, elf) {
		k = append(k, i)
	}
	i = Pos{
		x: pos.x - 1,
		y: pos.y,
	}
	if g.IsEnemy(i, elf) {
		k = append(k, i)
	}
	i = Pos{
		x: pos.x + 1,
		y: pos.y,
	}
	if g.IsEnemy(i, elf) {
		k = append(k, i)
	}
	i = Pos{
		x: pos.x,
		y: pos.y + 1,
	}
	if g.IsEnemy(i, elf) {
		k = append(k, i)
	}
	return k
}

type (
	Score struct {
		g int
		h int
	}

	PosScore struct {
		pos   Pos
		score Score
	}

	PosHeap struct {
		start  Pos
		list   []Pos
		scores map[Pos]Score
	}

	PosSet map[Pos]struct{}
)

func (s Score) f() int {
	return s.g + s.h
}

func NewPosHeap(start Pos) *PosHeap {
	return &PosHeap{
		start:  start,
		list:   []Pos{},
		scores: map[Pos]Score{},
	}
}

func (h PosHeap) Len() int {
	return len(h.list)
}
func (h PosHeap) Less(i, j int) bool {
	a := h.list[i]
	b := h.list[j]
	as := h.scores[a]
	bs := h.scores[b]
	af := as.f()
	bf := bs.f()
	if af == bf {
		if as.h == bs.h {
			return a.Less(b)
		}
		return as.h < bs.h
	}
	return af < bf
}
func (h PosHeap) Swap(i, j int) {
	h.list[i], h.list[j] = h.list[j], h.list[i]
}
func (h *PosHeap) Push(x interface{}) {
	h.list = append(h.list, x.(Pos))
}
func (h *PosHeap) Pop() interface{} {
	l := len(h.list)
	k := h.list[l-1]
	h.list = h.list[0 : l-1]
	return k
}
func (h *PosHeap) Add(g int, pos Pos) {
	h.scores[pos] = Score{
		g: g,
		h: 0,
	}
	heap.Push(h, pos)
}
func (h *PosHeap) Take() (Pos, int) {
	if h.Len() == 0 {
		return Pos{}, -1
	}
	k := heap.Pop(h).(Pos)
	ks := h.scores[k].g
	delete(h.scores, k)
	return k, ks
}
func (h *PosHeap) Has(pos Pos) bool {
	_, ok := h.scores[pos]
	return ok
}

func (ps PosSet) Has(pos Pos) bool {
	_, ok := ps[pos]
	return ok
}
func (ps PosSet) Add(pos Pos) {
	ps[pos] = struct{}{}
}

func (g *Game) RemoveEntity(e *Entity) {
	if e.elf {
		delete(g.elfs, e.pos)
		g.elfDied = true
	} else {
		delete(g.goblins, e.pos)
	}
}

func (g *Game) EntityPath(start Pos, goal PosSet) (Pos, int) {
	closed := PosSet{}
	open := NewPosHeap(start)
	for k, _ := range goal {
		open.Add(0, k)
	}
	for current, currentg := open.Take(); currentg > -1; current, currentg = open.Take() {
		if start.Manhattan(current) < 2 {
			return current, currentg + 1
		}
		closed.Add(current)
		k := g.AdjacentFree(current)
		for _, i := range k {
			if !closed.Has(i) {
				if !open.Has(i) {
					open.Add(currentg+1, i)
				} else if currentg+1 < open.scores[i].g {
					for n, j := range open.list {
						if i == j {
							node := open.scores[i]
							node.g = currentg + 1
							open.scores[i] = node
							heap.Fix(open, n)
							break
						}
					}
				}
			}
		}
	}
	return Pos{}, -1
}

func (g *Game) EntityMove(e *Entity, enemies map[Pos]*Entity) {
	inRange := PosSet{}
	for k, _ := range enemies {
		for _, i := range g.AdjacentFree(k) {
			inRange.Add(i)
		}
	}

	next, cost := g.EntityPath(e.pos, inRange)
	if cost < 0 {
		return
	}

	if e.elf {
		delete(g.elfs, e.pos)
		g.elfs[next] = e
	} else {
		delete(g.goblins, e.pos)
		g.goblins[next] = e
	}
	e.pos = next
}

func (g *Game) EntityAttack(e *Entity, enemies []*Entity) {
	if len(enemies) == 0 {
		return
	}
	target := enemies[0]
	for _, i := range enemies[1:] {
		if i.health < target.health {
			target = i
		}
	}
	if dead := e.Attack(target); dead {
		g.RemoveEntity(target)
	}
}

func (g *Game) TickEntity(e *Entity, enemies map[Pos]*Entity) {
	adjacentEnemies := g.AdjacentEnemy(e.pos, e.elf)
	if len(adjacentEnemies) == 0 {
		g.EntityMove(e, enemies)
		adjacentEnemies = g.AdjacentEnemy(e.pos, e.elf)
	}
	adj := make([]*Entity, 0, len(adjacentEnemies))
	for _, i := range adjacentEnemies {
		adj = append(adj, enemies[i])
	}
	g.EntityAttack(e, adj)
}

func (g *Game) Tick(part2 bool) (bool, bool) {
	all := make([]*Entity, 0, len(g.elfs)+len(g.goblins))
	for _, v := range g.elfs {
		all = append(all, v)
	}
	for _, v := range g.goblins {
		all = append(all, v)
	}
	sort.Sort(EntityByPos(all))
	for _, i := range all {
		if i.Dead() {
			continue
		}

		if len(g.elfs) == 0 || len(g.goblins) == 0 {
			return true, false
		}

		if i.elf {
			g.TickEntity(i, g.goblins)
		} else {
			g.TickEntity(i, g.elfs)
		}
		if part2 && g.elfDied {
			return false, true
		}
	}
	return false, false
}

func (g *Game) Print() {
	board := make([][]byte, len(g.board))
	for i := range board {
		k := make([]byte, len(g.board[i]))
		copy(k, g.board[i])
		board[i] = k
	}
	for k := range g.elfs {
		board[k.y][k.x] = 'E'
	}
	for k := range g.goblins {
		board[k.y][k.x] = 'G'
	}
	fmt.Print("   ")
	for i := 0; i < len(board[0]); i++ {
		fmt.Print(i % 10)
	}
	fmt.Println()
	for n, i := range board {
		fmt.Printf("%2d ", n)
		fmt.Println(string(i))
	}
}

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	board := [][]byte{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		board = append(board, []byte(scanner.Text()))
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	elfs := map[Pos]*Entity{}
	goblins := map[Pos]*Entity{}

	for y, row := range board {
		for x, i := range row {
			switch i {
			case byte('E'):
				pos := Pos{
					x: x,
					y: y,
				}
				elfs[pos] = NewEntity(pos, true)
				board[y][x] = '.'
			case byte('G'):
				pos := Pos{
					x: x,
					y: y,
				}
				k := NewEntity(pos, false)
				goblins[pos] = k
				board[y][x] = '.'
			}
		}
	}

	debug := false

	game := NewGame(copyMap(elfs), copyMap(goblins), board)

	i := 0
	for done, _ := game.Tick(false); !done; done, _ = game.Tick(false) {
		i++
		if debug {
			fmt.Println(i)
			game.Print()
		}
	}
	totalHealth := 0
	for _, i := range game.elfs {
		totalHealth += i.health
	}
	for _, i := range game.goblins {
		totalHealth += i.health
	}
	fmt.Println("rounds:", i)
	fmt.Println(i * totalHealth)

	part2(elfs, goblins, board)
}

func part2(elfs, goblins map[Pos]*Entity, board [][]byte) {
	for attackVal := 4; ; attackVal++ {
		elfNum := len(elfs)
		e := copyMap(elfs)
		for _, v := range e {
			v.attack = attackVal
		}
		game := NewGame(e, copyMap(goblins), board)
		i := 0
		for done, elfDied := game.Tick(true); !done && !elfDied; done, elfDied = game.Tick(true) {
			i++
		}
		if len(game.elfs) != elfNum {
			continue
		}
		totalHealth := 0
		for _, i := range game.elfs {
			totalHealth += i.health
		}
		for _, i := range game.goblins {
			totalHealth += i.health
		}
		fmt.Println("attack:", attackVal)
		fmt.Println(i * totalHealth)
		break
	}
}

File: day15/src/main.rs
use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap, HashSet};
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn distance(a: usize, b: usize) -> usize {
    if a > b {
        a - b
    } else {
        b - a
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Pos(usize, usize);

impl Pos {
    fn up(&self) -> Self {
        Pos(self.0, self.1 - 1)
    }
    fn down(&self) -> Self {
        Pos(self.0, self.1 + 1)
    }
    fn left(&self) -> Self {
        Pos(self.0 - 1, self.1)
    }
    fn right(&self) -> Self {
        Pos(self.0 + 1, self.1)
    }
    fn manhattan(&self, other: &Pos) -> usize {
        distance(self.0, other.0) + distance(self.1, other.1)
    }
}

impl Ord for Pos {
    fn cmp(&self, other: &Pos) -> Ordering {
        match self.1.cmp(&other.1) {
            Ordering::Equal => self.0.cmp(&other.0),
            i => i,
        }
    }
}

impl PartialOrd for Pos {
    fn partial_cmp(&self, other: &Pos) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

#[derive(Debug, Eq)]
struct Score {
    pos: Pos,
    g: usize,
}

impl Score {
    fn new(pos: Pos, g: usize) -> Self {
        Self { pos: pos, g: g }
    }
}

impl Ord for Score {
    fn cmp(&self, other: &Score) -> Ordering {
        match self.g.cmp(&other.g) {
            Ordering::Equal => self.pos.cmp(&other.pos).reverse(),
            i => i.reverse(),
        }
    }
}

impl PartialOrd for Score {
    fn partial_cmp(&self, other: &Score) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl PartialEq for Score {
    fn eq(&self, other: &Score) -> bool {
        self.pos == other.pos
    }
}

#[derive(Clone)]
struct Entity {
    pos: Pos,
    elf: bool,
    health: isize,
    attack: isize,
}

impl Entity {
    fn new(pos: Pos, elf: bool) -> Self {
        Self {
            pos: pos,
            elf: elf,
            health: 200,
            attack: 3,
        }
    }

    fn dead(&self) -> bool {
        self.health < 1
    }

    fn damage(&mut self, attack: isize) -> bool {
        self.health -= attack;
        self.dead()
    }
}

struct Game<'a> {
    elfs: HashMap<Pos, Entity>,
    goblins: HashMap<Pos, Entity>,
    board: &'a Vec<Vec<char>>,
}

impl<'a> Game<'a> {
    fn new(
        elfs: HashMap<Pos, Entity>,
        goblins: HashMap<Pos, Entity>,
        board: &'a Vec<Vec<char>>,
    ) -> Self {
        Self {
            elfs: elfs,
            goblins: goblins,
            board: board,
        }
    }

    fn adjacent_enemies(&self, pos: Pos, elf: bool) -> Vec<(Pos, isize)> {
        let enemies = if elf { &self.goblins } else { &self.elfs };
        let mut adjacent = Vec::new();
        if let Some(i) = enemies.get(&pos.up()) {
            adjacent.push((pos.up(), i.health));
        }
        if let Some(i) = enemies.get(&pos.left()) {
            adjacent.push((pos.left(), i.health));
        }
        if let Some(i) = enemies.get(&pos.right()) {
            adjacent.push((pos.right(), i.health));
        }
        if let Some(i) = enemies.get(&pos.down()) {
            adjacent.push((pos.down(), i.health));
        }
        adjacent
    }

    fn is_free(&self, pos: Pos) -> bool {
        !self.elfs.get(&pos).is_some()
            && !self.goblins.get(&pos).is_some()
            && self.board[pos.1][pos.0] != '#'
    }

    fn adjacent_free(&self, pos: Pos) -> Vec<Pos> {
        let mut adjacent = Vec::new();
        if self.is_free(pos.up()) {
            adjacent.push(pos.up());
        }
        if self.is_free(pos.left()) {
            adjacent.push(pos.left());
        }
        if self.is_free(pos.right()) {
            adjacent.push(pos.right());
        }
        if self.is_free(pos.down()) {
            adjacent.push(pos.down());
        }
        adjacent
    }

    fn path(&self, start: Pos, goal: HashSet<Pos>) -> Pos {
        let mut closed = HashSet::new();
        let mut openset = HashSet::new();
        let mut open = BinaryHeap::new();
        for i in goal.into_iter() {
            open.push(Score::new(i, 0));
        }
        while let Some(k) = open.pop() {
            if start.manhattan(&k.pos) < 2 {
                return k.pos;
            }
            openset.remove(&k.pos);
            closed.insert(k.pos);
            for i in self.adjacent_free(k.pos) {
                if !closed.contains(&i) && !openset.contains(&i) {
                    open.push(Score::new(i, k.g + 1));
                    openset.insert(i);
                }
            }
        }
        start
    }

    fn move_entity(&self, pos: Pos, elf: bool) -> Pos {
        let enemies = if elf { &self.goblins } else { &self.elfs };
        let mut in_range = HashSet::new();
        for &p in enemies.keys() {
            for i in self.adjacent_free(p) {
                in_range.insert(i);
            }
        }
        self.path(pos, in_range)
    }

    fn tick_entity(&self, e: &Entity) -> (Pos, Option<Pos>) {
        let mut adjacent = self.adjacent_enemies(e.pos, e.elf);
        let mut next = e.pos;
        if adjacent.len() == 0 {
            next = self.move_entity(e.pos, e.elf);
            adjacent = self.adjacent_enemies(next, e.elf);
        }
        (
            next,
            match adjacent.iter().min_by_key(|i| i.1) {
                Some(i) => Some(i.0),
                None => None,
            },
        )
    }

    fn tick(&mut self) -> bool {
        let mut all = self
            .elfs
            .values()
            .chain(self.goblins.values())
            .map(|i| (i.elf, i.pos))
            .collect::<Vec<_>>();
        all.sort_by_key(|i| i.1);
        for (is_elf, i) in all.into_iter() {
            let entity = match if is_elf {
                self.elfs.get(&i)
            } else {
                self.goblins.get(&i)
            } {
                Some(i) => i,
                None => continue,
            };
            if self.elfs.len() == 0 || self.goblins.len() == 0 {
                return false;
            }

            let (next, target) = self.tick_entity(entity);
            let attack = if is_elf {
                let mut entity = self.elfs.remove(&i).expect("Failed to get elf");
                let a = entity.attack;
                entity.pos = next;
                self.elfs.insert(next, entity);
                a
            } else {
                let mut entity = self.goblins.remove(&i).expect("Failed to get goblin");
                let a = entity.attack;
                entity.pos = next;
                self.goblins.insert(next, entity);
                a
            };
            match target {
                Some(k) => {
                    let enemy = if is_elf {
                        self.goblins
                            .get_mut(&k)
                            .expect("Failed to get enemy goblin")
                    } else {
                        self.elfs.get_mut(&k).expect("Failed to get enemy elf")
                    };
                    if enemy.damage(attack) {
                        if is_elf {
                            self.goblins.remove(&k).expect("Failed to remove goblin");
                        } else {
                            self.elfs.remove(&k).expect("Failed to remove elf");
                        }
                    }
                }
                None => (),
            }
        }
        true
    }

    fn print(&self) {
        let mut b = self.board.clone();
        for k in self.elfs.keys() {
            b[k.1][k.0] = 'E';
        }
        for k in self.goblins.keys() {
            b[k.1][k.0] = 'G';
        }
        for i in b.iter() {
            println!("{}", i.iter().collect::<String>());
        }
    }
}

fn main() {
    let file = File::open(PUZZLEINPUT).expect("Failed to open file");
    let reader = BufReader::new(file);

    let (elfs, goblins, board) = {
        let mut elfs = HashMap::new();
        let mut goblins = HashMap::new();
        let mut board = Vec::new();
        for (y, line) in reader.lines().enumerate() {
            let mut k = line
                .expect("Failed to read line")
                .chars()
                .collect::<Vec<_>>();
            for (x, i) in k.iter_mut().enumerate() {
                match i {
                    'E' => {
                        elfs.insert(Pos(x, y), Entity::new(Pos(x, y), true));
                        *i = '.';
                    }
                    'G' => {
                        goblins.insert(Pos(x, y), Entity::new(Pos(x, y), false));
                        *i = '.';
                    }
                    _ => (),
                }
            }
            board.push(k);
        }
        (elfs, goblins, board)
    };

    let mut game = Game::new(elfs.clone(), goblins.clone(), &board);
    let mut round = 0;
    while game.tick() {
        round += 1;
    }
    game.print();
    let total_health = game.elfs.values().map(|i| i.health).sum::<isize>()
        + game.goblins.values().map(|i| i.health).sum::<isize>();
    println!("{}", round * total_health);
}
